### PR TITLE
SDK feedback part 2

### PR DESCRIPTION
**Pull request by EvgeniyGushchin**
- set default limit value to 10 for TNEntityDiscoverBrandParams
- add FacetParameters to the SearchParams
- add facets params to search request
- Apply builder pattern to TNEntitySDKOptions
- add searchOptions property to search params
- move TNEntitySearchOptionIntent to TNEntitySearchOptions
- move TNEntitySDKOptions to separate file
- update folders structure
- add `validation` method to **Params** objects
- fix misspelling

### CODE REVIEW

File: TelenavEntitySDK/Classes/Configuration/TNEntitySDKOptions.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    public override init() {}
+    
+    public init(apiKey: String, apiSecret: String, deviceId: String? = nil, userId: String? = nil, locale: String? = nil, cloudEndPoint: String? = nil) {
+        self.apiKey = apiKey
+        self.apiSecret = apiSecret
+        self.deviceId = deviceId
+        self.userId = userId
+        self.locale = locale
+        self.cloudEndPoint = cloudEndPoint
+    }
```
Comment (can-wang) [inline TelenavEntitySDK/Classes/Configuration/TNEntitySDKOptions.swift:28]: Suggest to only support Builder() pattern here and get rid of the init(apiKey....) method.

File: TelenavEntitySDK/Classes/Model/Address&Coordinates/TNEntityGeoPoint.swift [hunk 2]
```diff
@@ -20,4 +20,8 @@ import Foundation
         let ret = TNEntityGeoPoint(lat: lat, lon: lon)
         return ret
     }
+    
+    public func isValid() -> Bool {
+        return latitude != nil && longitude != nil
```
Comment (can-wang) [inline TelenavEntitySDK/Classes/Model/Address&Coordinates/TNEntityGeoPoint.swift:25]: Please enhance the validation rules like below:
latitude != nil & latitude in [-90, 90]   && longitude != nill && longitude in [-180, 180]

**Review by can-wang — CHANGES_REQUESTED** [review]

File: TelenavEntitySDK/Classes/Validator.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        return $0.isEmpty ? "Must not be empty" : nil
+    })
+    
+    static let validScheme = Rule {
+        return $0.hasPrefix("http:") || $0.hasPrefix("https:") ? nil : "Incorrect scheme"
+    }
+    
+    static let validEntryTime = Rule {
+        let regex = #"^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2})(?:Z|([+-]\d{4})|)$"#
+
```
[the remark this answers is not in the document]
Comment (EvgeniyGushchin) [reply TelenavEntitySDK/Classes/Validator.swift:38]: @can-wang if we convert Date to string with this pattern "yyyy-MM-dd'T'HH:mmZ" then we will have something like `2020-12-28T17:17+0600` where +0600 is information about timezone
So I think that my regex is correct.

[the remark this answers is not in the document]
Comment (can-wang) [reply TelenavEntitySDK/Classes/Validator.swift:38]: Search REST API only accepts the entry time in two formats:
a. yyyy-MM-ddTHH:mmZ:   time in UTC time zone
b. yyy-MM-ddTHH:mm:  time in local time zone

Format like: 2020-12-12T12:10+0800 is not supported by search API, thus the using above regrex is incorect.

How about changing the entryTime in iOS SDK from string to Date and  convert it to string value in local time format "yyyy-MM-ddTHH:mm" when building the HTTP payload?

File: TelenavEntitySDK/Classes/Service/TNEntitySearchService.swift [hunk 4]
```diff
@@ -73,12 +73,13 @@ class TNEntitySearchService {
...
             }
             
-            params["search_options"] = ["show_address_lines" : searchParams.showAddressLines]
-            
+            if let facets = searchParams.facetParameters {
+                params["facet_parameters"] = facets.snakeCaseDictionary
+            }
         } else {
             
             let err = NSError(domain: "", code: 1001, userInfo: ["Description": "Both parameters searchParams and pageContext cannot be nil simultaneously"])
```
Comment (can-wang) [inline TelenavEntitySDK/Classes/Service/TNEntitySearchService.swift:85]: Should it here return TNEntityError ??

File: TelenavEntitySDK/Helpers/TNEntityError.swift [hunk 5]
```diff
@@ -10,26 +10,58 @@ import UIKit
...
+                return 1002
+            case .customMessage(_):
+                return 1003
+            }
+        }
+        
+        var message: String {
+            switch self {
+            case .authenticationError:
+                return "Invalid api key or api signature"
```
Comment (can-wang) [inline TelenavEntitySDK/Helpers/TNEntityError.swift:57]: Pls refine the message as "Invalid API key or signature"

File: TelenavEntitySDK/Helpers/TNEntityError.swift [hunk 6]
```diff
@@ -10,26 +10,58 @@ import UIKit
...
+    /// - Parameter type: TNErrorType
+    public init(_ type: TNErrorType) {
+        status = type.statusCode
+        message = type.message
+    }
+    
+    public enum TNErrorType {
+        case authenticationError
+        case parametersValidationError
+        case initializationError
```
Comment (can-wang) [inline TelenavEntitySDK/Helpers/TNEntityError.swift:38]: The name of initializationError is not very clear. 
Please rename it to  "ClientNotReadyError"

File: TelenavEntitySDK/Helpers/TNEntityError.swift [hunk 7]
```diff
@@ -10,26 +10,58 @@ import UIKit
...
+        case customMessage(message: String)
```
Comment (can-wang) [inline TelenavEntitySDK/Helpers/TNEntityError.swift:39]: "customMessage" is also not good name for error type.
Please rename it as "serviceError"

File: TelenavEntitySDK/Helpers/TNEntityError.swift [hunk 8]
```diff
@@ -10,26 +10,58 @@ import UIKit
...
+        
+        var statusCode: Int {
+            switch self {
+            case .authenticationError:
+                return 18401
```
Comment (can-wang) [inline TelenavEntitySDK/Helpers/TNEntityError.swift:44]: please change the code from 18401 to 1000

**Review by can-wang — CHANGES_REQUESTED** [review]

**Pull request merged**